### PR TITLE
[model] Add support for MiniCPM-V-4.6 model

### PR DESCRIPTION
**Pull request by tsjyma**
# PR type
- [ ] Bug Fix
- [ ] New Feature
- [ ] Document Updates
- More Models or Datasets Support

# PR information

## What
Add support for MiniCPM-V-4.6 model.

## Changes
- Add `minicpmv4_6` model type and template type
- Register model `OpenBMB/MiniCPM-V-4_6` with architecture `MiniCPMV4_6ForConditionalGeneration`
- Update `model_arch` for MiniCPM-V to support multiple sub-model keys
- Rework `MiniCPMV2Loader` to support `AutoModelForImageTextToText` / `AutoModelForVision2Seq`
- Add `_ensure_hf_device_map_for_meta` and `_materialize_tied_lm_head` helpers
- Add `MiniCPMV4_6Template` with NaViT-style vision tower support and packing/padding-free mode

### CODE REVIEW

File: swift/template/templates/minicpm.py [hunk 1]
```diff
@@ -604,3 +657,208 @@ def _data_collator(self, batch: List[Dict[str, Any]], *, padding_to: Optional[in
...
+            if valid_pvs and image_token_id is not None:
+                combined_pv = torch.cat(valid_pvs, dim=-1)
+                combined_ts = torch.cat(valid_ts, dim=0)
+                vision_output = model.model.get_image_features(
+                    combined_pv,
+                    combined_ts,
+                    downsample_mode=getattr(self, 'downsample_mode', None),
+                )
+                features = torch.cat(
+                    vision_output.pooler_output, dim=0).to(
```
[the remark this answers is not in the document]
Comment (tsjyma) [reply swift/template/templates/minicpm.py:659]: For minicpm-v-4.6, the processor will guarantee the image resolution is unified (normalized). So there're no such worries. You can refer to transformers v5.7.0 for minicpm-v-4.6 implementation.

File: swift/template/templates/minicpm.py [hunk 2]
```diff
@@ -604,3 +657,208 @@ def _data_collator(self, batch: List[Dict[str, Any]], *, padding_to: Optional[in
...
+
+    def _data_collator(self, batch, *, padding_to=None):
+        """Collect per-sample pixel_values / tgt_sizes intact (no dim-0 flatten)
+        so that _post_encode can concatenate them along the NaViT axis."""
+        res = {}
+        for k in ['pixel_values', 'tgt_sizes']:
+            items = [b.pop(k) for b in batch if b.get(k) is not None]
+            if items:
+                res[k] = items
+        res.update(Template._data_collator(self, batch, padding_to=padding_to))
```
[the remark this answers is not in the document]
Comment (tsjyma) [reply swift/template/templates/minicpm.py:624]: image_bound is replaced by get_placeholder_mask() for minicpm-v-4.6, so I need such implements

File: swift/model/patcher.py [hunk 3]
```diff
@@ -115,6 +115,17 @@ def _output_to_input_device_hook(module, args, kwargs, output):
 
 @contextmanager
 def patch_device_map():
+
+    def _ensure_no_split_modules():
+        for module in PreTrainedModel.__subclasses__():
+            if getattr(module, '_no_split_modules', None) is None:
+                module._no_split_modules = []
```
[the remark this answers is not in the document]
Comment (tsjyma) [reply swift/model/patcher.py]: This will be modified in following commit.

File: swift/model/models/minicpm.py [hunk 4]
```diff
@@ -53,16 +53,55 @@ def _get_vision_embedding(self, pixel_values):
...
+    if getattr(model, 'hf_device_map', None) is not None:
+        return
+    if not any(getattr(param, 'is_meta', False) for param in model.parameters()):
+        return
+    devices = {str(param.device) for param in model.parameters() if not getattr(param, 'is_meta', False)}
+    if len(devices) != 1:
+        return
+    device = next(iter(devices))
+    # Accelerate only treats the model as device-mapped when hf_device_map has more than one entry.
+    model.hf_device_map = {'llm': device, 'vision_tower': device}
```
[the remark this answers is not in the document]
Comment (tsjyma) [reply swift/model/models/minicpm.py:68]: Checked accelerate source, keys are only placeholders — accelerate never dispatches by hf_device_map keys. Keeping the generic names avoids version-specific coupling across v2.5/v2.6/v4/v4.5/v4.6. However, I did find and fix a related issue: model_arch.py's vision_tower list for minicpmv was missing 'model.vision_tower' (v4.6's actual attribute name), which would affect freeze_vit / vit_lr / LoRA targeting.

**Comment by Jintao-Huang** [comment]
Hello, I couldn't find this model.

**Comment by tsjyma** [comment]
Hello!  Currently, the weight of this new model is not released. But you can already find the implement of this model in transformers v5.7.0 https://github.com/huggingface/transformers/tree/main/src/transformers/models/minicpmv4_6.

**Comment by Jintao-Huang** [comment]
ok thanks!

**Comment by Jintao-Huang** [comment]
please run:
```
pip install pre-commit
pre-commit run --all-files
```

**Comment by Jintao-Huang** [comment]
Hello! Could you use inheritance as much as possible and avoid modifying the existing model code? That makes it harder to test.

**Review by Jintao-Huang — APPROVED** [review]

**Comment by tsjyma** [comment]
Thanks for the review! I've refactored the PR to use inheritance as suggested.

The key changes:

1. Created MiniCPMV4_6Loader(MiniCPMV2Loader) — all v4.6-specific model
   loading logic (AutoModelForImageTextToText, merger handling, tied lm_head
   materialization, meta device support) now lives exclusively in this subclass.

2. Reverted modifications to existing loaders:
   - MiniCPMVLoader.get_model — back to original (resampler.to, unconditional
     use_submodel_func('llm'))
   - MiniCPMV2Loader.get_model — back to original (patch_device_map +
     patch_output_clone only)
   - _patch_minicpmv_device_map — back to original (no merger branch)

3. Reverted model_arch.py — minicpmv arch keys restored to original single
   strings. Existing minicpmv/v2_5/v2_6/v4/v4_5 models are completely unaffected.

4. Reverted MiniCPMVTemplate._post_encode — back to original simple
   get_vllm_embedding call. The v4.6 template (MiniCPMV4_6Template) already
   overrides _post_encode with its own NaViT-style logic.

5. Moved downsample_mode from MiniCPMV2_6Template to MiniCPMV4_6Template
   (only v4.6 uses it).

The existing model/template code is now untouched — the v4.6 support is
fully contained in the new subclass layer.

File: swift/model/patcher.py [hunk 5]
```diff
@@ -115,6 +115,25 @@ def _output_to_input_device_hook(module, args, kwargs, output):
 
 @contextmanager
 def patch_device_map():
+
+    def _ensure_no_split_modules():
```
Comment (Jintao-Huang) [inline swift/model/patcher.py:119]: hello! Why was this logic added?

**Comment by Jintao-Huang** [comment]
Thanks for the PR, if there are no other changes, it will be merged today.

File: swift/model/patcher.py (hunk 5 above)
Comment (tsjyma) [reply swift/model/patcher.py:119]: Hello! The `_ensure_no_split_modules()` in `patch_device_map()` is a defensive guard
for device map inference during model loading.

When accelerate computes device_map="auto", it iterates over PreTrainedModel
subclasses and accesses their `_no_split_modules` attribute. If a subclass
(e.g. a nested language model within a VLM) has `_no_split_modules = None`,
accelerate raises an error.

This happens in two code paths:

1. Old transformers (no `_get_no_split_modules` method on PreTrainedModel):
   we can't monkey-patch, so we pre-populate all subclasses with `[]` once
   before yielding.

2. New transformers: we patch `_get_no_split_modules` to dynamically fill
   `_no_split_modules = []` for any module instance that has `None` at
   device-map computation time.

The MiniCPM-V-4.6 model architecture (with nested PreTrainedModel submodules
like `model.language_model`, `model.vpm`, `model.merger`) triggers this
because not all nested classes explicitly declare `_no_split_modules`.
This is a backward-compatible utility change that doesn't affect existing
model behavior — it only prevents a crash path that was already latent.

**Comment by Jintao-Huang** [comment]
https://github.com/modelscope/ms-swift/pull/9326
